### PR TITLE
planner: check whether plan-cache is disabled each time when getting plan from the plan cache

### LINKED ISSUE #49344: panics when set tidb_enable_prepared_plan_cache back to 0
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```
> create table t(a int);
> set @@tidb_enable_prepared_plan_cache=1;
> prepare s from "select * from t";
> set @@tidb_enable_prepared_plan_cache=0;
> execute s;
```

### 2. What did you expect to see? (Required)

execute should success

### 3. What did you see instead (Required)

```
TiDB root@127.0.0.1:test> execute s;
(1105, 'runtime error: invalid memory address or nil pointer dereference')
```

### 4. What is your TiDB version? (Required)

```
TiDB root@127.0.0.1:test> select tidb_version();
+-----------------------------------------------------------+
| tidb_version()                                            |
+-----------------------------------------------------------+
| Release Version: v7.6.0-alpha                             |
| Edition: Community                                        |
| Git Commit Hash: e62699cb26d6c6b98a52e6e998298dd86c7cdcb7 |
| Git Branch: heads/refs/tags/v7.6.0-alpha                  |
| UTC Build Time: 2023-12-10 14:24:33                       |
| GoVersion: go1.21.5                                       |
| Race Enabled: false                                       |
| Check Table Before Drop: false                            |
| Store: tikv                                               |
+-----------------------------------------------------------+
1 row in set
Time: 0.002s
```

### PR DESCRIPTION
**Pull request by qw4990**
### What problem does this PR solve?

Issue Number: close #49344

Problem Summary: planner: check whether plan-cache is disabled each time when getting plan from the plan cache

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/planner/core/plan_cache.go [hunk 1]
```diff
@@ -154,13 +154,16 @@ func GetPlanFromSessionPlanCache(ctx context.Context, sctx sessionctx.Context,
...
 	if isNonPrepared {
 		stmtCtx.CacheType = stmtctx.SessionNonPrepared
+		cacheEnabled = sctx.GetSessionVars().EnablePreparedPlanCache // plan-cache might be disabled after prepare.
 	} else {
 		stmtCtx.CacheType = stmtctx.SessionPrepared
+		cacheEnabled = sctx.GetSessionVars().EnableNonPreparedPlanCache
 	}
-	if !stmt.StmtCacheable {
+	stmtCtx.UseCache = stmt.StmtCacheable && cacheEnabled
+	if !stmt.StmtCacheable && stmt.UncacheableReason != "" {
```
Comment (AilinKid) [inline pkg/planner/core/plan_cache.go:166]: this `stmt.UncacheableReason != ""` is for what？

Comment (qw4990) [reply pkg/planner/core/plan_cache.go:166]: Just to ignore the empty warning, if the user disables the plan-cache, then `!StmtCacheable` is true, then the customer might receive an empty warning.

**Review by elsa0520 — APPROVED** [review]
LGTM

**Review by AilinKid — APPROVED** [review]

File: pkg/planner/core/plan_cache.go (hunk 1 above)
Comment (AilinKid) [reply pkg/planner/core/plan_cache.go:166]: make sense

**Pull request merged**